### PR TITLE
Delete project for `superuser`

### LINKED ISSUE #757: Delete a project (as superuser)
### User story

As a Lightning superuser...

1. I would like to be able to delete a project from the admin settings > projects page, so that I can remove projects that are no longer in use.
2. I would like to be able to cancel a project deletion, so that I can prevent a project from being deleted if I change my mind.
3. I would like to purge a project instantly, once it's been scheduled for deletion.

### Details

- The "delete" button (1) appears only for projects that have not been scheduled for deletion
- The "cancel deletion" (2) and "delete now" (3) buttons  appears only for projects that _HAVE_ been scheduled for deletion.

Change the projects page to use the same table component as the users page. Add the columns project name, scheduled for deletion and action  (delete) 
 
- When the user clicks delete, schedule the project for deletion `{PURGE_DELETED_RECORDS_AFTER_DAYS}` days from the current date and disable all of its jobs. 

The email functionality will be added in the owner issue

### Implementation notes

image

### User acceptance criteria
- [ ] on the projects page, there is a delete button
- [ ]  when the superuser clicks delete AND delete now, they get a confirmation modal before they can delete it
- [ ] when a user deletes a project (not delete_now), it is only scheduled for deletion. ie, the project is still there for PURGE_DELETED_RECORDS_AFTER_DAYS days
- [ ] when a user clicks 'delete_now', the project IS deleted and no longer present in the database. 
 - [ ] if a user inputs the wrong project name into the confirmation modal, they are unable to delete the project

### PR DESCRIPTION
**Pull request by mtuchi**
## Implementation plan
- [x] Create your new column  `scheduled_deletion`
- [x] use `Application.get_env(:lightning, :purge_deleted_after_days)` to set the scheduled deletion date when that “Delete” link is clicked by the superuser on the user's list page.
- [x] Clear the scheduled_deletion value when the “Cancel Delete” link is clicked by the super
- [x] Call `delete_project` when the superuser clicks the “Delete Now” link on that page.
- [x] Filter project that are scheduled for deletion
- [x] configure oban worker
- [x] fix failing tests

## Notes for the reviewer 
One of the tests is commented out because it depends on the [PR #759](https://github.com/OpenFn/Lightning/pull/798/files)
 [See the commented test ](https://github.com/OpenFn/Lightning/pull/806/files), I have also created this issue https://github.com/OpenFn/Lightning/issues/809 for updating project permission to follow new pattern 

To test the oban worker, edit the PURGE_AFTER_DELETED in runtime.exs:
```
config :lightning,
       :purge_deleted_after_days,
       System.get_env("PURGE_DELETED_AFTER_DAYS", "0")
       |> String.to_integer()
```

## Related issue

Fixes #757 

## Checklist before requesting a review

- [x] I have performed a **self-review** of my code
- [x] I have verified that all appropriate **authorization policies** have been implemented and tested
- [x] If needed, I have updated the **changelog**
- [x] Amber has **QA'd** this feature

### CODE REVIEW

File: lib/lightning_web/live/project_live/index.ex [hunk 1]
```diff
@@ -52,9 +52,97 @@ defmodule LightningWeb.ProjectLive.Index do
...
+           socket
+           |> assign(
+             active_menu_item: :projects,
+             projects: Projects.list_projects(),
+             page_title: "Projects"
+           )
+           |> put_flash(:info, "Project deleted successfully")
+           |> push_patch(to: ~p"/settings/projects")}
+
+        {:error, changeset} ->
```
Comment (mtuchi) [inline lib/lightning_web/live/project_live/index.ex:77]: @stuartc this is the line that codecov says i haven't covered in my tests

File: lib/lightning_web/live/project_live/index.ex [hunk 2]
```diff
@@ -52,9 +52,97 @@ defmodule LightningWeb.ProjectLive.Index do
...
+    project = Projects.get_project(project_id)
+
+    case Projects.cancel_scheduled_deletion(project) do
+      {:ok, _project} ->
+        {:noreply,
+         socket
+         |> put_flash(:info, "Canceled project deletion schedule")
+         |> push_patch(to: ~p"/settings/projects")}
+
+      {:error, %Ecto.Changeset{} = changeset} ->
```
Comment (mtuchi) [inline lib/lightning_web/live/project_live/index.ex:108]: @stuartc this is  another line

File: lib/lightning_web/live/project_live/index.ex [hunk 3]
```diff
@@ -52,9 +52,97 @@ defmodule LightningWeb.ProjectLive.Index do
...
+    project = Projects.get_project(project_id)
+
+    case Projects.schedule_project_deletion(project) do
+      {:ok, _project} ->
+        {:noreply,
+         socket
+         |> put_flash(:info, "Project scheduled for deletion")
+         |> push_patch(to: ~p"/settings/projects")}
+
+      {:error, %Ecto.Changeset{} = changeset} ->
```
Comment (mtuchi) [inline lib/lightning_web/live/project_live/index.ex:127]: @stuartc another line

Comment (stuartc) [reply lib/lightning_web/live/project_live/index.ex:127]: Ok I've looked into all of these, they are either very difficult to test because we can't really control the conditions in order to trigger a failure; or they have no validations like the one above this one and technically could never return an error.

I don't think these matches should be removed though.

File: lib/lightning_web/live/project_live/index.ex [hunk 4]
```diff
@@ -52,9 +52,97 @@ defmodule LightningWeb.ProjectLive.Index do
...
+
   def can_delete_project(current_user, project),
-    do: ProjectUsers |> Permissions.can?(:delete_project, current_user, project)
+    do: Users |> Permissions.can?(:delete_project, current_user, project)
+
+  def request_project_deletion(current_user, project) do
+    is_nil(project.scheduled_deletion) and
+      (can_request_delete_project(current_user, project) or
+         current_user.role == :superuser)
+  end
```
Comment (stuartc) [inline lib/lightning_web/live/project_live/index.ex:147]: Why are we checking for `:superuser` here? All of the permission logic should be inside the policies.
The `:request_delete_project` action/permission should be able to handle the superuser case.

**Review by stuartc — CHANGES_REQUESTED** [review]
So re: the test coverage, it's fine. I'm not that worried for now, we don't have a lot of control over the sad path.

I've made two comments for you to address, but other than that this is looking good!

File: lib/lightning_web/live/project_live/index.ex (hunk 4 above)
Comment (elias-ba) [reply lib/lightning_web/live/project_live/index.ex]: Done

File: lib/lightning_web/live/components/deletion_modal.ex [hunk 5]
```diff
@@ -181,4 +233,68 @@ defmodule LightningWeb.Components.UserDeletionModal do
...
+          :let={f}
+          for={@scheduled_deletion_changeset}
+          phx-change="validate"
+          phx-submit="delete"
+          phx-target={@myself}
+          id="scheduled_deletion_form"
+        >
+          <span>
+            This project and all data associated to it (workflows, jobs, project users, project credentials, ...) will be deleted.
+          </span>
```
Comment (amberrignell) [inline lib/lightning_web/live/components/deletion_modal.ex:255]: "This project and all data associated to it (workflows, jobs, project users, project credentials, ...) will be deleted. Please make sure none of the workflows in it are still in use."

File: test/lightning/policies/project_user_permissions_test.exs [hunk 6]
```diff
@@ -56,6 +69,14 @@ defmodule Lightning.Policies.ProjectUserPermissionsTest do
       assert ProjectUsers |> Permissions.can?(:access_project, viewer, project)
     end
 
+    test "can not access that scheduled deletion project", %{
```
Comment (amberrignell) [inline test/lightning/policies/project_user_permissions_test.exs:72]: "can not access a project that is scheduled for deletion"

File: test/lightning/policies/project_user_permissions_test.exs [hunk 7]
```diff
@@ -146,14 +165,10 @@ defmodule Lightning.Policies.ProjectUserPermissionsTest do
...
     end
-
-    test "cannot delete the project", %{project: project, admin: admin} do
-      refute ProjectUsers |> Permissions.can?(:delete_project, admin, project)
-    end
   end
 
   describe "Project users with the :owner role" do
-    test "can create workflows, create / edit / delete / run / rerun jobs, edit the project name, edit the project description, and delete the project.",
+    test "can create workflows, create / edit / delete / run / rerun jobs, edit the project name, edit the project description, and request project deletion.",
```
Comment (amberrignell) [inline test/lightning/policies/project_user_permissions_test.exs:171]: is this "request project deletion" being tested?

**Review by amberrignell — CHANGES_REQUESTED** [review]

File: lib/lightning/projects.ex [hunk 8]
```diff
@@ -407,4 +419,39 @@ defmodule Lightning.Projects do
...
+    |> Enum.each(fn job ->
+      Lightning.Jobs.update_job(job, %{
+        "enabled" => false
+      })
+    end)
+
+    Project.scheduled_deletion_changeset(project, %{
+      "scheduled_deletion" => date
+    })
+    |> Repo.update()
```
[the remark this answers is not in the document]
Comment (elias-ba) [reply lib/lightning/projects.ex]: Done

File: lib/lightning_web/live/components/deletion_modal.ex (hunk 5 above)
Comment (elias-ba) [reply lib/lightning_web/live/components/deletion_modal.ex:255]: Done

File: test/lightning/policies/project_user_permissions_test.exs (hunk 6 above)
Comment (elias-ba) [reply test/lightning/policies/project_user_permissions_test.exs:72]: Done

File: test/lightning/policies/project_user_permissions_test.exs (hunk 7 above)
Comment (elias-ba) [reply test/lightning/policies/project_user_permissions_test.exs:171]: Done

File: lib/lightning/projects.ex [hunk 9]
```diff
@@ -340,7 +384,7 @@ defmodule Lightning.Projects do
   def first_project_for_user(user) do
```
[the remark this answers is not in the document]
Comment (elias-ba) [reply lib/lightning/projects.ex:384]: Done

File: lib/lightning_web/live/components/project_deletion_modal.ex [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+  @impl true
+  def handle_event(
+        "validate",
+        %{"project" => project_params},
+        socket
+      ) do
+    changeset =
+      socket.assigns.project
+      |> Projects.change_scheduled_deletion(project_params)
+      |> Map.put(:action, :validate_scheduled_deletion)
```
Comment (stuartc) [inline lib/lightning_web/live/components/project_deletion_modal.ex:30]: Just curious, why `:validation_scheduled_deletion`, normally we put `:validate` because of LiveView forms, do the errors appear correctly on the form if you don't enter the matching project name?

File: lib/lightning_web/live/components/project_deletion_modal.ex [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+
+  alias Lightning.Projects
+  alias Lightning.Projects.Project
+
+  @impl true
+  def update(%{project: project} = assigns, socket) do
+    {:ok,
+     socket
+     |> assign(
+       delete_now?: !is_nil(project.scheduled_deletion),
```
Comment (stuartc) [inline lib/lightning_web/live/components/project_deletion_modal.ex:15]: Perhaps it would be clearer if this was `:scheduled_for_deletion?`?

File: lib/lightning_web/live/project_live/form_component.ex [hunk 12]
```diff
@@ -184,7 +184,19 @@ defmodule LightningWeb.ProjectLive.FormComponent do
...
+       when not is_map_key(project_params, "project_users") do
+    []
+  end
+
+  defp filter_users_to_notify(
+         socket,
+         project_params
+       )
+       when is_map_key(project_params, "project_users") do
     project = Repo.preload(socket.assigns.project, :project_users)
```
Comment (stuartc) [inline lib/lightning_web/live/project_live/form_component.ex:200]: Why is this necessary? I don't see what else changed that makes this a requirement? Is there not some kind of `@target` issue here on the edit page perhaps?

**Review by stuartc — CHANGES_REQUESTED** [review]
Hey dude, I think there is a lot going on here, it's not a small feature - but there are a bunch of things that don't quite make sense to me.

Please give my comments a read, if you need some clarity as well please let me know.

File: lib/lightning_web/hooks.ex [hunk 13]
```diff
@@ -20,6 +20,7 @@ defmodule LightningWeb.Hooks do
 
     projects =
       Lightning.Projects.get_projects_for_user(socket.assigns.current_user)
+      |> Enum.filter(fn project -> is_nil(project.scheduled_deletion) end)
```
[the remark this answers is not in the document]
Comment (elias-ba) [reply lib/lightning_web/hooks.ex:23]: Done

**Comment by elias-ba** [comment]
Closing in favor of #847
